### PR TITLE
trie: add trie db wrapper; refactor trienode

### PR DESCRIPTION
**Pull request by Francesco4203**
References: 
- https://github.com/ethereum/go-ethereum/pull/27160 (only for moving memoryNode to NodeSet and rename) 
- https://github.com/ethereum/go-ethereum/pull/27176

Changes:
- Add wrapper for database since later, we will have another database for path-based. Note that the common features such as cache, preimage,... are in the wrapper interface.
- Move nodeset to trienode/node 
- Adjust tests: add scheme to tests for later testing on path-based
- Remove callback in db.Commit() since it is not used (ref https://github.com/ethereum/go-ethereum/pull/26637)

### CODE REVIEW

**Comment by huyngopt1994** [comment]
noted, it will split trie/database.go -> into trie/triedb/<hashdb/pathdb> structure.

File: trie/trienode/node.go [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+	"sort"
+	"strings"
+
+	"github.com/ethereum/go-ethereum/common"
+)
+
+// Node is a wrapper which contains the encoded blob of the trie node and its
+// unique hash identifier. It is general enough that can be used to represent
+// trie nodes corresponding to different trie implementations.
+type Node struct {
```
Comment (huyngopt1994) [inline trie/trienode/node.go:30]: Rename from memoryNode-> Node

File: trie/trienode/node.go [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+	return len(n.Blob) + common.HashLength
+}
+
+// IsDeleted returns the indicator if the node is marked as deleted.
+func (n *Node) IsDeleted() bool {
+	return n.Hash == (common.Hash{})
+}
+
+// WithPrev wraps the Node with the previous node value attached.
+type WithPrev struct {
```
Comment (huyngopt1994) [inline trie/trienode/node.go:46]: NodeWithPrev

File: trie/iterator_test.go [hunk 3]
```diff
@@ -612,5 +671,33 @@ func TestIteratorNodeBlob(t *testing.T) {
...
+// only single layer trie is used, namely storage trie is not
+// considered at all.
+func isTrieNode(scheme string, key, val []byte) (bool, []byte, common.Hash) {
+	var (
+		path []byte
+		hash common.Hash
+	)
+	if scheme == rawdb.HashScheme {
+		ok := rawdb.IsLegacyTrieNode(key, val)
+		if !ok {
```
Comment (huyngopt1994) [inline trie/iterator_test.go:690]: @Francesco4203 Could u add some unittest specific for IsLegacyTrieNode and IsAccountTrieNode?

File: trie/trienode/node.go [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+	if n.IsDeleted() {
+		set.deletes += 1
+	} else {
+		set.updates += 1
+	}
+	set.Nodes[string(path)] = n
+}
+
+// AddLeaf adds the provided leaf node into set.
+func (set *NodeSet) AddLeaf(parent common.Hash, blob []byte) {
```
Comment (huyngopt1994) [inline trie/trienode/node.go:125]: before, they init leaf obj and pass directly here. 
 ```// addLeaf collects the provided leaf node into set.
func (set *NodeSet) addLeaf(leaf *leaf) {
	set.leaves = append(set.leaves, leaf)
}

File: trie/database_test.go [hunk 5]
```diff
@@ -17,17 +17,20 @@
...
-	if _, err := db.Node(common.Hash{}); err == nil {
-		t.Fatalf("metaroot retrieval succeeded")
+// newTestDatabase initializes the trie database with specified scheme.
+
+func newTestDatabase(diskdb ethdb.Database, scheme string) *Database {
+	db := prepare(diskdb, nil)
+	if scheme == rawdb.HashScheme {
+		db.backend = hashdb.New(diskdb, db.cleans, mptResolver{})
 	}
+	//} else {
```
Comment (huyngopt1994) [inline trie/database_test.go:32]: hm Why we comment here. ?

Comment (Francesco4203) [reply trie/database_test.go:32]: It will be later enabled for path based scheme

File: trie/iterator_test.go (hunk 3 above)
Comment (Francesco4203) [reply trie/iterator_test.go:690]: Added

**Review by huyngopt1994 — APPROVED** [review]
LGTM

**Pull request merged**